src/qibo/derivative.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+def parameter_shift(
+    circuit, observable, parameter_index, gate_eigenv, initial_state=None
+):
+    """In this method the parameter shift rule (PSR) is implemented.
+    Given a circuit U and an observable H, the PSR allows to calculate the derivative
+    of the expected value of H on the final state obtained by application of U with
+    respect to a variational parameter of the circuit.
+
+    Original references:
+        `https://arxiv.org/abs/1811.11184`;
+        `https://arxiv.org/abs/1803.00745`.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
+        observable (:class: `qibo.hamiltonians.Hamiltonian`): the hamiltonian here is the target observable.
+        parameter_index (int): the index which identify the targe parameter in the circuit.get_parameters list
+        gate_eigenv (float): abs(eigenvalue) of H. In case of Pauli 1/2{sigmas} observable r = 0.5
+        initial_state ((1, 2**nqubits) matrix): initial state on which we act with the circuit.
+
+    Returns:
+        np.float value of the derivative of the expected value of H on the final state obtained applying U
+        to the initial state with respect to the target variational parameter.
+
+    Example:
+        .. testcode::
```

Comment:
I think this example is quite extensive here. We should certainly write a test using this approach (comparing with tensorflow) and maybe we could also use it in the "Advanced Examples" sections in the docs. For the docstring I would use a more minimal example that only shows how to use qibo's PSR, not the comparison with tensorflow.

Reply:
@stavros11 regarding the previous comment about duplication of arguments (I don't know why GitHub doesn't let me write the reply up there):

There is no such duplication between `observable` and `gate_eigenv`, the eigenvalue refers to the generator of the parametrized gate, which is completely independent of the observable measured at the end of the circuit.

However, it is true that there is some duplication: from the `parameter_index` we could get the type of gate the parameter belongs to, and from this get the eigenvalue of the generator of the gate. I guess this is what we should do, right?

Reply:
Moreover, the latter would also allow us to raise an `NotImplementedError` if the current parameter shift rule does not apply to the gate in question.